File: bindings/r/chemometrics4all/src/libc4a/src/core/augmentations/noise/gaussian_noise.c
/* SPDX-License-Identifier: CECILL-2.1 */
/*
 * GaussianAdditiveNoise — reference implementation.
 *
 * std-per-row computation matches NumPy's biased estimator (ddof=0):
 *
 *     mu_i  = (1/p) * sum_j X[i, j]
 *     var_i = (1/p) * sum_j (X[i, j] - mu_i)^2
 *     std_i = sqrt(var_i)
 *
 * which is the two-pass form NumPy uses for `np.std(X, axis=1)`. The single
 * `standard_normal_fill(rows*cols)` call produces the noise matrix in
 * row-major order; multiplying by `std_i * sigma` yields the per-sample
 * noise term and the final `out[i, j] = X[i, j] + noise[i, j]`.
 */

#include "gaussian_noise.h"

#include <math.h>
#include <stdlib.h>

struct c4a_aug_gaussian_noise_state_t {
    c4a_rng_pcg64* rng;   /* borrowed; not owned. */
    double sigma;
};

c4a_aug_gaussian_noise_state_t* c4a_aug_gaussian_noise_state_new(
    c4a_rng_pcg64* rng, double sigma) {
    c4a_aug_gaussian_noise_state_t* s =
        (c4a_aug_gaussian_noise_state_t*)malloc(sizeof(*s));
    if (s == NULL) return NULL;
    s->rng   = rng;
    s->sigma = sigma;
    return s;
}

void c4a_aug_gaussian_noise_state_free(c4a_aug_gaussian_noise_state_t* state) {
    free(state);
}

c4a_status_t c4a_aug_gaussian_noise_state_apply(
    c4a_aug_gaussian_noise_state_t* state,
    const double* X, int64_t rows, int64_t cols, double* out) {
    if (state == NULL || state->rng == NULL) return C4A_ERR_NULL_POINTER;
    if (X == NULL || out == NULL)             return C4A_ERR_NULL_POINTER;
    if (rows < 0 || cols < 0)                 return C4A_ERR_INVALID_ARGUMENT;
    if (rows == 0 || cols == 0)               return C4A_OK;

    const size_t n_total = (size_t)rows * (size_t)cols;

    /* 1. Per-row std (biased estimator, two-pass form). */
    double* stds = (double*)malloc((size_t)rows * sizeof(double));
    if (stds == NULL) return C4A_ERR_OUT_OF_MEMORY;
    for (int64_t i = 0; i < rows; ++i) {
        double mu = 0.0;
        const double* row = X + (size_t)i * (size_t)cols;
        for (int64_t j = 0; j < cols; ++j) {
            mu += row[(size_t)j];
        }
        mu /= (double)cols;
        double var = 0.0;
        for (int64_t j = 0; j < cols; ++j) {
            const double d = row[(size_t)j] - mu;
            var += d * d;
        }
        var /= (double)cols;
        stds[(size_t)i] = sqrt(var);
    }

    /* 2. Draw the noise matrix (row-major) with one bulk call. */
    double* noise = (double*)malloc(n_total * sizeof(double));
    if (noise == NULL) {
        free(stds);
        return C4A_ERR_OUT_OF_MEMORY;
    }
    c4a_pcg64_engine_standard_normal_fill(state->rng, noise, n_total);

    /* 3. out[i, j] = X[i, j] + noise[i, j] * stds[i] * sigma. */
    const double sigma = state->sigma;
    for (int64_t i = 0; i < rows; ++i) {
        const double scale = stds[(size_t)i] * sigma;
        const size_t off   = (size_t)i * (size_t)cols;
        for (int64_t j = 0; j < cols; ++j) {
            out[off + (size_t)j] =
                X[off + (size_t)j] + noise[off + (size_t)j] * scale;
        }
    }

    free(noise);
    free(stds);
    return C4A_OK;
}

File: parity/donor_imports/python_generator/src/n4m_parity_augmenters_ref/gaussian_noise.py
# SPDX-License-Identifier: CECILL-2.1
"""
GaussianAdditiveNoise — frozen NumPy reference.

Per-sample std-scaled Gaussian noise:

    stds[i]      = std(X[i, :])                  # biased estimator, ddof=0
    noise[i, j]  = standard_normal(rows * cols)  # row-major reshape
    out[i, j]    = X[i, j] + noise[i, j] * stds[i] * sigma

The noise matrix is drawn from one
``rng.standard_normal((rows, cols))`` call so the RNG stream consumed
matches the C side's single ``standard_normal_fill(rows * cols)``.
"""

from __future__ import annotations

import numpy as np


__all__ = ["gaussian_noise_apply"]


def gaussian_noise_apply(
    X: np.ndarray, *, rng: np.random.Generator, sigma: float
) -> np.ndarray:
    """Apply additive Gaussian noise to spectra.

    Parameters
    ----------
    X : np.ndarray, shape (n_samples, n_features), dtype float64
        Input spectra (row-major).
    rng : np.random.Generator
        PCG64 generator (seeded by the caller).
    sigma : float
        Relative noise scale; the per-sample std is multiplied by sigma to
        get the actual noise stddev.

    Returns
    -------
    out : np.ndarray, same shape and dtype as X
    """
    X = np.ascontiguousarray(X, dtype=np.float64)
    if X.ndim != 2:
        raise ValueError(f"X must be 2-D; got {X.ndim}-D")
    n_samples, n_features = X.shape
    if n_samples == 0 or n_features == 0:
        return X.copy()

    # Per-sample std, biased (NumPy default ddof=0).
    stds = np.std(X, axis=1)  # shape (n_samples,)
    # One bulk draw, reshape row-major.
    noise = rng.standard_normal(size=n_samples * n_features).reshape(
        n_samples, n_features
    )
    scale = (stds * sigma).reshape(-1, 1)
    return X + noise * scale
